File: src/algorithms/python/Count_Primes.py
# -*- coding: utf-8 -*-

from math import sqrt


class Solution:
    # @param {integer} n
    # @return {integer}
    def countPrimes(self, n):
        if n < 2:
            return 0
        e = int(sqrt(n)) + 1
        vis, ret = [False for i in range(n)], n - 2
        vis[0], vis[1] = True, True
        for num in range(2, n, 2):
            vis[num] = True
        if n > 2:
            ret = n / 2
        for num in range(3, e, 2):
            vis[num] = True
            for _num in range(num * num, n, num * 2):
                if not vis[_num]:
                    ret -= 1
                    vis[_num] = True
        return ret


if __name__ == "__main__":
    solution = Solution()
    print solution.countPrimes(2)
    print solution.countPrimes(10)
    print solution.countPrimes(10000)

File: src/algorithms/cpp/Count_Primes.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int countPrimes(int n) {
        if (n < 2)  return 0;
        int e = int(sqrt(n)) + 1;
        vector <bool> vis (n, false);
        int result = n - 2;
        vis[0] = true;
        vis[1] = true;
        for (int i = 4; i < n; i += 2) {
            vis[i] = true;
            result -= 1;
        }
        for (int i = 3; i < e; i += 2) {
            for (int j = i * i; j < n; j += i * 2) {
                if (!vis[j])    result -= 1, vis[j] = true;
            }
        }
        return result;
    }
};

int main() {
    return 0;
}
